libevmasm/RuleList.h

```diff
@@ -629,6 +630,18 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart7(
 		feasibilityFunction
 	});
 
+	rules.push_back({
+		// AND(SHL(A,X), SHL(A,Y)) -> SHL(A, AND(X,Y))
+		Builtins::AND(Builtins::SHL(A,X), Builtins::SHL(A,Y)),
+		[=]() -> Pattern { return Builtins::SHL(A, Builtins::AND(X,Y)); }
+	});
+
+	rules.push_back({
+		// AND(SHL(Z,X), SHL(Z,Y)) -> SHL(Z, AND(X,Y))
```

Comment:
Please note that there is a huge difference between A B C and X Y Z. A B C are always known constants, X Y Z are arbitrary complex expressions that don't need to be constant.

Please replace `A` above by `Z` and compare gas costs.

Reply:
If I replace `A` with `Z`, the rule at `RuleList.h: 633` becomes exactly the same as the rule at `RuleList.h:639`